tools/junitreport/pkg/parser/gotest/data_parser.go

```diff
@@ -0,0 +1,142 @@
+func (p *testDataParser) ExtractMessage(line string) (string, bool) {
+	// `go test` doesn't have messages
+	return "", false
+}
+
+// MarksCompletion determines if the line marks the completion of a test case
+func (p *testDataParser) MarksCompletion(line string) bool {
+	// output of a test case follows the result line for some indefinite amount of lines in `go test` output
+	// so we can't pin down which line marks the completion of a test case
+	return false
+}
+
+func newTestSuiteDataParser() parser.TestSuiteDataParser {
+	return &testSuiteDataParser{
+		// coverageOutputPattern matches coverage output on a single line.
+		// The first submatch of this regex is the percent coverage
+		coverageOutputPattern: regexp.MustCompile(`coverage:\s+(\d+\.\d+)\% of statements`),
+
+		// packageResultPattern matches the `go test` output for the end of a package.
+		// The first submatch of this regex matches the result of the test (ok or FAIL)
+		// The second submatch of this regex matches the name of the package
+		// The third submatch of this regex matches the time taken in seconds for tests in the package to finish
+		// The sixth (optional) submatch of this regex is the percent coverage
+		packageResultPattern: regexp.MustCompile(`(ok|FAIL)\s+(.+)[\s\t]+(\d+\.\d+(s| seconds))([\s\t]+coverage:\s+(\d+\.\d+)\% of statements)?`),
+	}
+}
+
+type testSuiteDataParser struct {
+	coverageOutputPattern *regexp.Regexp
+	packageResultPattern  *regexp.Regexp
+}
+
+// MarksBeginning determines if the line marks the begining of a test suite
+func (p *testSuiteDataParser) MarksBeginning(line string) bool {
+	// `go test` doesn't mark the beginning of a package
+	return false
+}
+
+// ExtractName extracts the name of the test suite from a test output line
+func (p *testSuiteDataParser) ExtractName(line string) (name string, ok bool) {
+	if matches := p.packageResultPattern.FindStringSubmatch(line); len(matches) > 1 && len(matches[2]) > 0 {
+		return matches[2], true
+	}
+	return "", false
+}
+
+const (
+	coveragePropertyName string = "coverage.statements.pct"
+)
+
+// ExtractProperties extracts any metadata properties of the test suite from a test output line
+func (p *testSuiteDataParser) ExtractProperties(line string) (properties map[string]string, ok bool) {
+	// the only test suite properties that Go testing can create are coverage values, which can either
+	// be present on their own line or in the package result line
+	if matches := p.coverageOutputPattern.FindStringSubmatch(line); len(matches) > 0 && len(matches[1]) > 0 {
+		return map[string]string{
+			coveragePropertyName: matches[1],
+		}, true
+	}
+
+	if resultMatches := p.packageResultPattern.FindStringSubmatch(line); len(resultMatches) > 5 && len(resultMatches[6]) > 0 {
+		return map[string]string{
+			coveragePropertyName: resultMatches[6],
+		}, true
+	}
+	return map[string]string{}, false
+}
```

Comment:
Duration isn't a property, but a first-class field on the package. This will take a larger refactor. I don't think my interfaces are worth it - there are too many differences between `go test` output and others, adding an `ExtractDuration` method to the package data extractor would not be a good idea - output of `test/cmd` for instance won't use it.